src/Features/Core/Portable/Diagnostics/EngineV2/CompilerDiagnosticExecutor.cs

```diff
@@ -0,0 +1,292 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.EngineV2
+{
+    /// <summary>
+    /// Diagnostic Executor that only relies on compiler layer. this might be replaced by new CompilationWithAnalyzer API.
+    /// </summary>
+    internal static class CompilerDiagnosticExecutor
+    {
+        public static async Task<ImmutableDictionary<DiagnosticAnalyzer, AnalysisResult>> AnalyzeAsync(this CompilationWithAnalyzers analyzerDriver, Project project, CancellationToken cancellationToken)
+        {
+            var version = await DiagnosticIncrementalAnalyzer.GetDiagnosticVersionAsync(project, cancellationToken).ConfigureAwait(false);
+
+            // Run all analyzers at once.
+            // REVIEW: why there are 2 different cancellation token? one that I can give to constructor and one I can give in to each method?
+            // REVIEW: we drop all those allocations for the diagnostics returned. can we avoid this?
+            await analyzerDriver.GetAnalyzerDiagnosticsAsync(cancellationToken).ConfigureAwait(false);
+
+            // this is wierd, but now we iterate through each analyzer for each tree to get cached result.
+            // REVIEW: no better way to do this?
+            var noSpanFilter = default(TextSpan?);
+            var analyzers = analyzerDriver.Analyzers;
+            var compilation = analyzerDriver.Compilation;
+
+            var builder = ImmutableDictionary.CreateBuilder<DiagnosticAnalyzer, AnalysisResult>();
+            foreach (var analyzer in analyzers)
+            {
+                var result = new Builder(project, version);
+
+                // REVIEW: more unnecessary allocations just to get diagnostics per analyzer
+                var oneAnalyzers = ImmutableArray.Create(analyzer);
+
+                foreach (var tree in compilation.SyntaxTrees)
+                {
+                    var model = compilation.GetSemanticModel(tree);
+
+                    var syntax = await analyzerDriver.GetAnalyzerSyntaxDiagnosticsAsync(tree, oneAnalyzers, cancellationToken).ConfigureAwait(false);
+                    result.AddSyntaxDiagnostics(tree, CompilationWithAnalyzers.GetEffectiveDiagnostics(syntax, compilation));
```

Comment:
You needn't have to invoke GetEffectiveDiagnostics, all GetAnalyzerXXXDiagnosticsAsync APIs should return effective diagnostics. Did you hit a case where this is not true?